mlflow/litellm/__init__.py

```diff
@@ -0,0 +1,144 @@
+
+            # NB: We don't need to patch async function because Databricks notebook waits
+            # for the async task to finish before finishing the cell.
+    else:
+        litellm.success_callback = _remove_mlflow_callbacks(litellm.success_callback)
+        litellm.failure_callback = _remove_mlflow_callbacks(litellm.failure_callback)
+        # Callback also needs to be removed from 'callbacks' as litellm adds
+        # success/failure callbacks to there as well.
+        litellm.callbacks = _remove_mlflow_callbacks(litellm.callbacks)
+
+
+# NB: The @autologging_integration annotation must be applied here, and the callback injection
+# needs to happen outside the annotated function. This is because the annotated function is NOT
+# executed when disable=True is passed. This prevents us from removing our callback and patching
+# when autologging is turned off.
+@autologging_integration(FLAVOR_NAME)
+def _autolog(
+    log_traces: bool,
+    disable: bool = False,
+    silent: bool = False,
+):
+    pass
+
+
+def _patch_theading_in_function(target, function_name: str):
+    """
+    Apply the threading patch to a synchronous function.
+
+    We capture the threads started by the function using the _patch_thread_start context manager,
+    then join them to ensure they are finished before the notebook cell finishes executing.
+    """
+
+    def _patch_fn(original, *args, **kwargs):
+        with _patch_thread_start() as logging_threads:
+            result = original(*args, **kwargs)
+        for thread in logging_threads:
+            thread.join()
+        return result
+
+    safe_patch(FLAVOR_NAME, target, function_name, _patch_fn)
+
+
+@contextmanager
+def _patch_thread_start():
+    """
+    A context manager to collect threads started for logging handlers.
+    This is done by monkey-patching the start() method of threading.Thread.
+    Note that failure handlers are executed synchronously, so we don't need to patch them.
+    """
+    original = Thread.start
+    logging_threads = []
+
+    def patched_thread(self, *args, **kwargs):
+        target = getattr(self, "_target", None)
+        # success_handler is for normal request, and run_success_... is for streaming
+        # - https://github.com/BerriAI/litellm/blob/4f8a3fd4cfc20cf43b38379928b41c2691c85d36/litellm/utils.py#L946
+        # - https://github.com/BerriAI/litellm/blob/4f8a3fd4cfc20cf43b38379928b41c2691c85d36/litellm/utils.py#L7526
+        if target and target.__name__ in [
+            "success_handler",
+            "run_success_logging_and_cache_storage",
+        ]:
+            logging_threads.append(self)
+        return original(self, *args, **kwargs)
+
+    Thread.start = patched_thread
+    try:
+        yield logging_threads
+    finally:
+        Thread.start = original
+
+
+def _append_mlflow_callbacks(original_callbacks):
+    if not any(cb == "mlflow" for cb in original_callbacks):
```

Comment:
Is it possible original_callbacks is None?

Reply:
Not really - they are initialized with an empty list: https://github.com/BerriAI/litellm/blob/70aa85af1fc01d30934ee71a703b8d3982420d05/litellm/__init__.py#L39-L42